### DISCUSSION TITLE
framerate during video

### DISCUSSION THREAD

**Discussion by BUDMAN65** [raspberrypi/picamera2#111] [Q&A]
Can anyone give an example of setting the framerate during video recording?

**Comment by davidplowman** [comment]
After calling `picam2.configure(config)` and before `picam2.start()` use `picam2.set_controls({"FrameDurationLimits": (33333, 33333)})`.

The two numbers are the maximum and minimum frame times in usecs. So this example would give you fixed 30fps.

**Reply by andrewsg3** [reply]
Do you find this actually works? My setup is Pi 5 with Pi HQ camera. Using frame duration limits seems to have some ability to set framerate within a window (10-50fps) but caps out around 48fps. It cannot reach anywhere near 120fps like rpicam-vid can. This is even with main size set to 240p.

**Reply by davidplowman** [reply]
> Do you find this actually works? My setup is Pi 5 with Pi HQ camera. Using frame duration limits seems to have some ability to set framerate within a window (10-50fps) but caps out around 48fps. It cannot reach anywhere near 120fps like rpicam-vid can. This is even with main size set to 240p.

Please include the code you are using that caps out at 48fps. Be sure to delete any code that is not related to the video recording so that the test case is as small as possible. Thank you.

**Reply by andrewsg3** [reply]
Hi,

I will preface this by saying I am relatively new to computer vision, so I'm assuming this is user error.

I am using a Pi 5 with HQ camera. I have set the main size according to the output of rpicam-hello, given here:

```
(env) pi@mosdacpi:~/Desktop $ rpicam-hello --list-cameras
Available cameras
-----------------
0 : imx477 [4056x3040 12-bit RGGB] (/base/axi/pcie@1000120000/rp1/i2c@88000/imx477@1a)
    Modes: 'SRGGB10_CSI2P' : 1332x990 [120.05 fps - (696, 528)/2664x1980 crop]
           'SRGGB12_CSI2P' : 2028x1080 [50.03 fps - (0, 440)/4056x2160 crop]
                             2028x1520 [40.01 fps - (0, 0)/4056x3040 crop]
                             4056x3040 [10.00 fps - (0, 0)/4056x3040 crop]
```

So image size 1332x990 should achieve 120fps. Indeed, rpicam-vid seems able to achieve 120fps video when running

`rpicam-vid --framerate 120 --nopreview -o "rpicam_vid_120fps.mp4`

Shown here in a video with a counter incrementing at 120Hz:
https://github.com/user-attachments/assets/74789032-6351-43a0-830b-ed8efed1b886

With ffprobe confirming:
![image](https://github.com/user-attachments/assets/e18730ba-cb2b-4a5c-a268-9f3d541ca33a)

Now, here is a picamera2 example I set up where I attempt to record at 120fps. 

```
import picamera2
import time

picam2 = picamera2.Picamera2()

target_fps = 120

frame_limits = int(1000000/target_fps)

print(f"Target fps = {target_fps}, frame duration limits = {frame_limits}, {frame_limits}")

config = picam2.create_video_configuration(
	main={"size": (1332,990)},
	controls={"FrameDurationLimits": [frame_limits, frame_limits]}
)

picam2.configure(config)

encoder = picamera2.encoders.H264Encoder(bitrate=1000000)

filename = f"picamera2_{target_fps}fps.mp4"
output = picamera2.outputs.FileOutput(filename)

picam2.start_recording(encoder, output)

time.sleep(5)

picam2.stop_recording()
```

The video can be played with ffplay, but does not open with vlc or microsoft videos app (after exporting to my PC):
https://github.com/user-attachments/assets/c4d7442e-b1a0-4d5f-bdbb-3c8e353e71cb

And ffprobe reports 30fps:
![image](https://github.com/user-attachments/assets/c353bf0c-bf2a-471e-960a-7d2070bd3965)

It's probably something to do with how I'm setting up my encoder and FileOutput object, but it's beyond me at this stage.

I tried exporting as .h264, incase it was something to do with the .mp4 extension - similar result

ffprobe:
![image](https://github.com/user-attachments/assets/13bfd9bf-c210-4f57-a3be-cdbabb4ab00c)

video: I can't upload .h264 to github, so I used ffmpeg to convert it to .mp4 with:

`ffmpeg -framerate 120 -i picamera2_120fps.h264 -c copy picamera2_120fps_h264tomp4.mp4`

which is here: 
https://github.com/user-attachments/assets/47347c0f-6eb4-4af8-8f22-1d4bdadf1b0d

but ffprobe still reports 30fps:
![image](https://github.com/user-attachments/assets/28d31c97-aad5-4067-bf84-0e80687fbdf2)

Any advice is much appreciated.

Edit: correct code tags. It's three backquotes in a line of their own to start and end a code block.

**Reply by andrewsg3** [reply]
I thought it might also be helpful to see this output during recording, rpicam-vid and picamera2 show different stream configurations and sensor modes - this seems a likely culprit?

rpicam-vid
![image](https://github.com/user-attachments/assets/6b942cc9-52b7-474c-89b6-e61f0aa66bd7)

picamera2
![image](https://github.com/user-attachments/assets/87b09026-0ac2-494f-b868-0041c8c4ad44)

**Reply by andrewsg3** [reply]
Following the guidance in 4.2.2.3 of the picamera2 manual I have attempted to manually configure sensor mode

```
import picamera2
import time

picam2 = picamera2.Picamera2()

mode = picam2.sensor_modes[0]

print(f"Selected mode has size {mode['size']}, bit depth {mode['bit_depth']} and fps {mode['fps']}.")

config = picam2.create_video_configuration(sensor={'output_size': mode['size'], 'bit_depth':
mode['bit_depth']})

picam2.configure(config)

target_fps = 120

frame_limits = int(1000000/target_fps)

print(f"Target fps = {target_fps}, frame duration limits = {frame_limits}, {frame_limits}")

config = picam2.create_video_configuration(
        main={"size": (1332,990)},
        controls={"FrameDurationLimits": [frame_limits, frame_limits]}
)

picam2.configure(config)

encoder = picamera2.encoders.H264Encoder(bitrate=1000000)

filename = f"picamera2_{target_fps}fps.mp4"
output = picamera2.outputs.FileOutput(filename)

picam2.start_recording(encoder, output)

time.sleep(5)

picam2.stop_recording()
```

Again ffprobe reports 30fps, video is slow, and the sensor configuration does not match rpicam-vid
![image](https://github.com/user-attachments/assets/e4f7e404-36f9-440d-97b8-7c95fe15e00a)

Edit: more code tags fixed

**Reply by davidplowman** [reply]
Hi, so it looks like you've more or less figured everything out, you've got it to choose the right camera mode (often a stumbling block!) and so on.

But the obvious problem is that you're using the `FileOutput`. This writes an unencapsulated raw H.264 bistream to a file. It so happens you're calling it ".mp4" but that doesn't make it an mp4 file! Raw H.264 bitstreams don't contain timestamps, and any software reading it (if it _can_ read it) may just default to 30fps.

You need to choose an output that can write a real mp4 file, like this:
```
import time
from picamera2 import Picamera2
from picamera2.encoders import H264Encoder
from picamera2.outputs import PyavOutput

picam2 = Picamera2()

mode = picam2.sensor_modes[0]

sensor = {'output_size': mode['size'], 'bit_depth': mode['bit_depth']}

main = {'size': (1332, 990)}
controls = {'FrameRate': 120}
config = picam2.create_video_configuration(main, sensor=sensor, controls=controls)
picam2.configure(config)

encoder = H264Encoder(bitrate=10000000)
output = PyavOutput("test.mp4")

picam2.start_recording(encoder, output)

time.sleep(5)

picam2.stop_recording()
```
I'm getting about 118fps here according to `ffprobe`. Note that 1332x990@120fps is probably slightly exceeding the system's capabilities. We quote 1080p30 as being supported, but note that a Pi 5 should normally get up to about 1080p50. In terms of pixel throughput, that's about the same as 720p100 so the limit will be somewhere around there. For me, 720p120 seems to work OK, but it doesn't quite keep up at 1332x990@120, but it's very close.

**Reply by andrewsg3** [reply]
Thank you so much, been able to achieve approximately 99fps with that code which is great.

**Reply by davidplowman** [reply]
A couple of other thoughts. If you set the main stream format to "YUV420" that may give you better performance. Also maybe try increasing `buffer_count` in `create_video_configuration`. Maybe try `buffer_count=16`.

**Reply by andrewsg3** [reply]
Yes! That bumped it to 119.44fps. You don't know how much stress this has saved me! Thanks again :)
